src/test/java/net/oauth/jsontoken/AbstractJsonTokenParserTest.java

```diff
@@ -45,91 +47,121 @@ public void testVerify_valid() throws Exception {
+
+    assertThrowsWithErrorCode(
+        IllegalStateException.class,
+        ErrorCode.EXPIRED_TOKEN,
+        () -> parser.verify(checkToken, getVerifiers())
+    );
   }
 
   public void testVerify_expirationNull() throws Exception {
     Instant issuedAt = clock.now().plus(SKEW.minus(Duration.standardSeconds(1)));
-    assertTrue(verifyTimeFrame(issuedAt, null));
+    AbstractJsonTokenParser parser = getAbstractJsonTokenParser();
+    JsonToken checkToken = getJsonTokenWithTimeRange(issuedAt, null);
+
+    parser.verify(checkToken, getVerifiers());
   }
 
   public void testVerify_issuedAtNullExpirationNull() throws Exception {
-    assertTrue(verifyTimeFrame(null, null));
-  }
+    AbstractJsonTokenParser parser = getAbstractJsonTokenParser();
+    JsonToken checkToken = getJsonTokenWithTimeRange(null, null);
 
-  public void testVerify_futureToken() throws Exception {
-    Instant issuedAt = clock.now().plus(SKEW.plus(Duration.standardSeconds(1)));
-    Instant expiration = issuedAt.plus(Duration.standardDays(1));
-    assertFalse(verifyTimeFrame(issuedAt, expiration));
+    parser.verify(checkToken, getVerifiers());
   }
 
-  public void testVerify_pastToken() throws Exception {
-    Instant expiration = clock.now().minus(SKEW.plus(Duration.standardSeconds(1)));
-    Instant issuedAt = expiration.minus(Duration.standardDays(1));
-    assertFalse(verifyTimeFrame(issuedAt, expiration));
-  }
 
   public void testVerify_badSignature() throws Exception {
     AbstractJsonTokenParser parser = getAbstractJsonTokenParser();
     JsonToken checkToken = naiveDeserialize(TOKEN_STRING_BAD_SIG);
-    assertThrows(
+    assertThrowsWithErrorCode(
         SignatureException.class,
+        ErrorCode.BAD_SIGNATURE,
         () -> parser.verify(checkToken, getVerifiers())
     );
   }
 
   public void testVerify_emptySignature() throws Exception {
     AbstractJsonTokenParser parser = getAbstractJsonTokenParser();
     JsonToken checkToken = naiveDeserialize(TOKEN_STRING_EMPTY_SIG);
-    assertThrows(
+    assertThrowsWithErrorCode(
         SignatureException.class,
+        ErrorCode.BAD_SIGNATURE,
         () -> parser.verify(checkToken, getVerifiers())
     );
   }
 
   public void testVerify_nullSignature() throws Exception {
     AbstractJsonTokenParser parser = getAbstractJsonTokenParser();
     JsonToken checkToken = naiveDeserialize(TOKEN_STRING_2PARTS);
-    assertThrows(
+    assertThrowsWithErrorCode(
         IllegalStateException.class,
+        ErrorCode.MALFORMED_TOKEN_STRING,
         () -> parser.verify(checkToken, getVerifiers())
     );
   }
 
   public void testVerify_unsupportedSignatureAlgorithm() throws Exception {
     AbstractJsonTokenParser parser = getAbstractJsonTokenParser();
     JsonToken checkToken = naiveDeserialize(TOKEN_STRING_UNSUPPORTED_SIGNATURE_ALGORITHM);
-    assertThrows(
+    // This function does not explicitly check or access the signature algorithm
```

Comment:
Do we need this test case? It's testing parser.verify in the same way as testVerify_emptyVerifiers right?

Reply:
The underlying logic being run is somewhat different. With empty verifiers, the function just immediately fails. However, with an unsupported signature algorithm, rather than checking that the signature algorithm is invalid/unsupported and throwing another error (to indicate that the algorithm is unsupported), the library still compares the base string and signature using the verifyFunction of each of the provided verifiers.Since this behavior was somewhat unintuitive, I added a test case to specifically document this behavior. 

The unsupported signature algorithm test case probably closer to the bad signature test case. Even with an unsupported algorithm in the header, if the signature itself is valid, then the verify function passes. (eg. Having HS512 as the alg in the header, but the signature is generated using HS256. This would pass if the HS256 verifier was inputted.)